fix(spark-connector): filter out all pg gravitino table properties

Xander-run opened the pull request:
<!--
1. Title: [#<issue>] <type>(<scope>): <subject>
   Examples:
     - "[#123] feat(operator): support xxx"
     - "[#233] fix: check null before access result in xxx"
     - "[MINOR] refactor: fix typo in variable name"
     - "[MINOR] docs: fix typo in README"
     - "[#255] test: fix flaky test NameOfTheTest"
   Reference: https://www.conventionalcommits.org/en/v1.0.0/
2. If the PR is unfinished, please mark this PR as draft.
-->

### What changes were proposed in this pull request?

(Please outline the changes and how this PR fixes the issue.)

### Why are the changes needed?

(Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, describe the bug.)

Fix: #(issue)

### Does this PR introduce _any_ user-facing change?

(Please list the user-facing changes introduced by your change, including
  1. Change in user-facing APIs.
  2. Addition or removal of property keys.)

### How was this patch tested?

(Please test your changes, and provide instructions on how to test it:
  1. If you add a feature or fix a bug, add a test to cover your changes.
  2. If you fix a flaky test, repeat it for many times to prove it works.)

Xander-run commented:
I'm having trouble running integration test locally (probably my laptop's performance issue). 

```
ava.util.concurrent.TimeoutException

            at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)

            at org.apache.gravitino.integration.test.MiniGravitino.start(MiniGravitino.java:170)

            ... 76 more
```

@FANNG1 Could you please kindly approve the integration test WF?

FANNG1 commented on spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/jdbc/postgersql/PostgreSqlPropertiesConverter.java:
```diff
@@ -0,0 +1,75 @@
+import org.apache.gravitino.spark.connector.PropertiesConverter;
+import org.apache.gravitino.spark.connector.jdbc.JdbcPropertiesConstants;
+
+public class PostgreSqlPropertiesConverter implements PropertiesConverter {
```
how about letting the pg converter extend `JdbcPropertiesConverter`? since more logics are same

Xander-run replied:
As the default constructor of JdbcPropertiesConverter is made private, it couldn't be extended.

FANNG1 replied:
We could add `getPGInstance` method in `PostgreSqlPropertiesConverter` to create a `JdbcPropertiesConverter ` which doesn't support table properties.

FANNG1 replied:
```
public class JdbcPropertiesConverter implements PropertiesConverter {

  private final boolean supportsTableProperties;

  public static class JdbcPropertiesConverterHolder {
    private static final JdbcPropertiesConverter INSTANCE = new JdbcPropertiesConverter(true);
    private static final JdbcPropertiesConverter PG_INSTANCE = new JdbcPropertiesConverter(false);
  }

  private JdbcPropertiesConverter(boolean supportsTableProperties) {
    this.supportsTableProperties = supportsTableProperties;
  }

  public static JdbcPropertiesConverter getPGInstance() {
    return JdbcPropertiesConverterHolder.PG_INSTANCE;
  }

  @Override
  public Map<String, String> toGravitinoTableProperties(Map<String, String> properties) {
   // do some logic according to `supportsTableProperties`
  }
  
}
```

FANNG1 commented on spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/jdbc/postgersql/PostgreSqlPropertiesConverter.java:
```diff
@@ -0,0 +1,75 @@
+
+  @Override
+  public Map<String, String> toGravitinoTableProperties(Map<String, String> properties) {
+    // PG doesn't support table properties.
```
How about throwing an exception if the properties contains the properties beside `owner`?

Xander-run replied:
I changed the logic to only filter out property "owner" and the rest unchanged. As property comment will be used to add comment to PG table.

FANNG1 commented on spark-connector/v3.5/spark/src/test/java/org/apache/gravitino/spark/connector/integration/test/jdbc/SparkJdbcMysqlCatalogIT35.java:
```diff
@@ -23,7 +23,7 @@
 import org.junit.jupiter.api.Test;
 
-public class SparkJdbcMysqlCatalogIT35 extends SparkJdbcMysqlCatalogIT {
+public class SparkJdbcMysqlCatalogIT35 extends SparkJdbcPostgreSqlCatalogIT {
```
seems not correct

Xander-run replied:
fixed

FANNG1 commented on spark-connector/spark-common/src/test/java/org/apache/gravitino/spark/connector/integration/test/jdbc/SparkJdbcPostgreSqlCatalogIT.java:
```diff
@@ -0,0 +1,108 @@
+  protected void initCatalogEnv() throws Exception {
+    ContainerSuite containerSuite = ContainerSuite.getInstance();
+    containerSuite.startPostgreSQLContainer(PG_CATALOG_PG_IT);
+    pgUrl = containerSuite.getPostgreSQLContainer().getJdbcUrl();
```
please add `this.` for pgXX varibles.

Xander-run replied:
fixed

Xander-run commented:
Due to the issue tracked in [apache/gravitino#6907](https://github.com/apache/gravitino/issues/6907), the testListTables and testListTable tests in SparkJdbcPostgreSqlCatalogIT are currently failing. As a temporary workaround, these tests can be disabled and marked as TODO until the issue is resolved.

FANNG1 commented on integration-test-common/src/test/java/org/apache/gravitino/integration/test/util/TestDatabaseName.java:
```diff
@@ -102,4 +102,12 @@ public String toString() {
       return this.name().toLowerCase();
     }
   },
+
```
Could you add a comment for `PG_CATALOG_PG_IT`?

FANNG1 commented on spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/jdbc/postgersql/PostgreSqlPropertiesConverter.java:
```diff
@@ -0,0 +1,80 @@
+        filteredProperties.put(key, properties.get(key));
+      }
+    }
+    return filteredProperties;
```
How about throwing an exception explicitly here?

Xander-run replied:
Thanks for the update! I'm a bit unclear about the reasoning behind throwing an exception here. If we do throw, wouldn't the SparkConnector still fail to create the PG table because of the exception? Are we aiming to explicitly block table creation in this case, or should we allow it to proceed by silently filtering out the unsupported 'owner' property?

FANNG1 replied:
If we check here, we could provide an meaningful error message and no need to pass to Gravitino or the background PG.

Xander-run replied:
Got it. I’ve disabled the Spark PG connector tests since table creation is not yet supported. Most test cases require table creation before the test functionalities can be executed.

FANNG1 commented:
> Due to the issue tracked in [apache/gravitino#6907](https://github.com/apache/gravitino/issues/6907), the `testListTables` and `testListTable` tests in `SparkJdbcPostgreSqlCatalogIT` are currently failing. As a temporary workaround, these tests can be disabled and marked as TODO until the issue is resolved.

suggest to add an SupportsXX method to condition execute the test.

FANNG1 commented:
@Xander-run this PR is planed to merge to 0.9 release, is there anything blocks you?

Xander-run commented:
> @Xander-run this PR is planed to merge to 0.9 release, is there anything blocks you?

Hi @FANNG1, thanks for the review! I've addressed all your comments. There's just one last thing that needs to be resolved in https://github.com/apache/gravitino/pull/6824#discussion_r2051776859. Then we're good to go.

FANNG1 commented on spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/jdbc/JdbcPropertiesConverter.java:
```diff
@@ -63,6 +76,13 @@ public Map<String, String> toSparkCatalogProperties(Map<String, String> properti
   public Map<String, String> toGravitinoTableProperties(Map<String, String> properties) {
+    if (!this.supportsTableProperties) {
+      for (String key : properties.keySet()) {
+        if (key.equals("owner")) {
```
How about using the following code?
```java
for (String key : properties.keySet()) {
  if (!"owner".equalsIgnoreCase(key)) {
     throw new UnsupportedOperationException("Doesn't support table property " + key);
  }
}
return  Collections.emptyMap();
```

FANNG1 commented on spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/jdbc/JdbcPropertiesConverter.java:
```diff
@@ -63,7 +76,16 @@ public Map<String, String> toSparkCatalogProperties(Map<String, String> properti
+          throw new UnsupportedOperationException("Doesn't support table property " + key);
+        }
+      }
+      return new HashMap<>();
```
Collections.emptyMap is preferred because it doesn't create an new map every time, but this is not an blocking issue, let's go ahead

FANNG1 commented:
@Xander-run , merged to main, thanks for your contribution!
